renderer/backend/GpuBuffer.cpp

```diff
@@ -0,0 +1,144 @@
+/*****************************************************************************
+                    The Dark Mod GPL Source Code
+
+ This file is part of the The Dark Mod Source Code, originally based
+ on the Doom 3 GPL Source Code as published in 2011.
+
+ The Dark Mod Source Code is free software: you can redistribute it
+ and/or modify it under the terms of the GNU General Public License as
+ published by the Free Software Foundation, either version 3 of the License,
+ or (at your option) any later version. For details, see LICENSE.TXT.
+
+ Project: The Dark Mod (http://www.thedarkmod.com/)
+
+******************************************************************************/
+#include "precompiled.h"
+#include "GpuBuffer.h"
+
+extern idCVarBool r_usePersistentMapping;
+
+const int GpuBuffer::NUM_FRAMES;
+
+void GpuBuffer::Init( GLenum type, GLuint size, GLuint alignment, byte *staticData, GLuint numStaticBytes ) {
+	assert( idMath::IsPowerOfTwo( alignment ) && "Alignment must be a power of 2" );
+	
+	if( bufferObject ) {
+		Destroy();
+	}
+
+	staticDataSize = ALIGN( numStaticBytes, alignment );
+	frameSize = ALIGN( size, alignment );
+	this->alignment = alignment;
+	this->type = type;
+
+	qglGenBuffers( 1, &bufferObject );
+	qglBindBuffer( type, bufferObject );
+
+	usesPersistentMapping = r_usePersistentMapping && GLAD_GL_ARB_buffer_storage;
+
+	totalSize = staticDataSize + NUM_FRAMES * frameSize;
+	if ( usesPersistentMapping ) {
+		qglBufferStorage( type, totalSize, nullptr, GL_MAP_WRITE_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT );
+		bufferContents = ( byte* )qglMapBufferRange( type, 0, totalSize, GL_MAP_WRITE_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT );
```

Comment:
I don't understand the technical details behind it, but yes, GL_MAP_COHERENT_BIT guarantees both.

For point 1, I've seen a bit of the workings due to Nvidia nSight which shows when it's copying. Copies happen right before the first draw call that's accessing state from that buffer. E.g. when you have a certain range of the buffer bound and you issue a draw call, it will trigger an upload. Due to the bound range, it also knows which regions it needs to copy. 
However, I don't know how it knows that the region is even dirty to begin with, but it does - subsequent draw calls don't retrigger the upload :)

Reply:
It's magic? =)

    That is, data written to the store by either the client or server will
    be visible to any subsequently issued GL commands with no further action
    taken by the application.

It seems that `visible to any subsequently issued GL commands` has replaced the initial wording `immediately visible` (Public Bug 935 in specs).

I see two ways it can be implemented:

 * **API way**. At the closest GL API call, the driver copies the whole mapped range. This is very bad because all three buffers are mapped, meaning that implementation can copy all of them every frame.

 * **Paging way**. The mapped memory pages are specifically marked, so that first write to them triggers page fault or similar exception. Driver catches the exception and marks the page as modified somewhere. On the first GL call it uploads all modified pages. This is pretty strange understanding, because it would mean cross-platform OpenGL standard requires a memory-mapping unit on host (maybe it is OK today). However, it also means that memory is uploaded in pages with at least 4KB alignment, so we should force-align `frameSize` to page size, because otherwise we can accidentally upload a bit of data which we can potentially be drawn by GPU right now.

Reply:
Moreover, there seems to be no way to learn of the page size, if paging is used in implementation. Which is even more weird. As for potential overwrite, perhaps driver simply allows it because logically we must not change the actual data with it.

Reply:
Again, no idea how they do it, but I'm reasonably certain the usage pattern I employ is correct and performant. It is essentially as proposed in the "approaching zero driver overhead" talks by the people who actually work on the GPU drivers, so I'd hope they know what they're doing :D 
But I do think the mapped buffer memory on the CPU side is some kind of specially mapped construct controlled by the driver, so I'd speculate that your "paging way" may touch on the actual implementation. But I think the control/tracking would have to be as fine as the minimum buffer alignment requirements, which I think would be 32 bytes, to guarantee correctness.

Reply:
This can't be "memory controlled by driver", this is memory directly accessed by CPU instructions. It has to use some mechanism supported by CPU.

I think a more likely possibility is using write-combined memory for mapped regions. This memory is uncached, but subsequent writes to such memory are hold in WC buffer in order to combine them to a larger write. If this memory region points directly to GPU memory, then it will make changes visible sort-of immediately. Maybe some flush is necessary to push the last bits, but it should be cheap to do at any API call. However, if there are separate memory buffers on CPU and on GPU, then it is unclear how could driver understand when to copy.

Reply:
Well, you clearly understand more of this than I do. Unfortunately, I didn't find any deeper technical explanation about how that COHERENT_BIT part is actually implemented.

My understanding of the techniques used in this PR mostly come from these two talks:
https://www.youtube.com/watch?v=K70QbvzB6II (AZDO: approaching zero driver overhead) and
https://www.youtube.com/watch?v=-bCeNzgiJ8I (Beyond Porting)
If you don't know them, you might find them interesting for insight into how persistent buffer, multidraw and bindless textures work together to eliminate driver overhead. But unfortunately neither of them goes into detail of how persistent buffers actually work :-/

Reply:
This post suggests that persistent buffers may reside in a shared memory region accessible to both CPU and GPU, so there may not actually be any memory transfer: https://community.khronos.org/t/persistent-buffer-mapping-what-kind-of-magic-is-this/71646

Reply:
Yes, looks like the memory is not duplicated, so the version of non-cacheable WC writes is the most likely. They get flushed with any CPU-serializing instruction like SFENCE or CPUID. It's quite likely that most of API calls do it with low performance cost.